### PR TITLE
[NF] Add Concatenate tracks workflows

### PR DESCRIPTION
**Pull request by skoudoro**
This PR is a follow up of #2760.  

It migrates the script `concatenate_tractogram` from TRX-Python library to DIPY.

More PRs should come to continues the migration.

### CODE REVIEW

**Comment by skoudoro** [comment]
@frheault can you have a look ?

**Comment by frheault** [comment]
That's great/useful to transfer! I haven't tested yet due to the fact I can't build Dipy right now.

Am I missing instructions since the new 3.12 update? I tried the good old `pip install -e .` and did not work so I tried:

```
1. pip install -r requirements/build.txt
2. pip install -e .
3. meson setup --reconfigure builddir/
4. meson -C compile builddir/
5. pip install -e .
```

And when I try to launch the script I get `FileNotFoundError: [Errno 2] No such file or directory: '/tmp/pip-build-env-ogw3qo_g/normal/local/bin/ninja'`. Am I missing a package? a `sudo apt install`? Or is there instructions I did not see?

When I got that figure out I will be able to run tests, try the script, etc. Also I think the script should be name **dipy_concatenate_tractograms**

**Comment by skoudoro** [comment]
```
pip install -r requirements/build.txt
pip install --no-build-isolation -e .
```

should be enough to install DIPY.

Concerning the name, I might rename it just `dipy_concatenate` and allow to concatenate tractograms or volumes

**Comment by frheault** [comment]
Thanks, I will try that!

About a script for both, it would be confusing to see parameters that applies to only tractograms or only volumes or some for both, no?

Like the datatype, the verification of header (can be turned off for tractography, but shape has to match for volume, or data_per_*, etc.) that would make the help twice as long and more confusing

File: dipy/workflows/io.py [hunk 1]
```diff
@@ -257,3 +261,81 @@ def run(self, input_files, vol_idx=0, out_dir='',
...
+                has_group = True
+            trx_list.append(tractogram_obj)
+
+        trx = concatenate_tractogram(
+            trx_list, delete_dpv=delete_dpv, delete_dps=delete_dps,
+            delete_groups=delete_groups or not has_group,
+            check_space_attributes=check_space_attributes,
+            preallocation=preallocation)
+
+        valid_extensions = ['trk', 'trx', "tck,", "fib", "dpy", "vtk"]
```
Comment (frheault) [inline dipy/workflows/io.py:335]: There is a comma in between the quote for tck

**Comment by skoudoro** [comment]
> About a script for both, it would be confusing to see parameters that applies to only tractograms or only volumes or some for both, no?

you are right. I was thinking in subcommand, like "dipy_concat. tractogram args" a bit like git which should avoid input parameter conflict/problem/confusion

But let's keep it simple for now.  I will update that. Thanks for the review

**Comment by skoudoro** [comment]
Hi @frheault,

Did you succeed to install DIPY?  Did you have time to try this PR? 

Thank you in advance for you feedback

**Comment by frheault** [comment]
Ah there is a missing comment, I forgot to click somewhere. 
Sorry about that.
I installed it correctly and tested the script. I tried 5 TRX/TRK/TCK -> TRX/TRK/TCK (all combinaison) 

The TRK had data_per_point and it was conserved correctly, also the header were alright and the position was matching the reference anatomy.

So it is working well!

**Comment by skoudoro** [comment]
Thank you for the feedback!

I will wait until Friday in case someone else want to try it and then, I will go ahead and merge it

**Comment by skoudoro** [comment]
working on `validate_tractogram` and `convert_tractogram`, I start to wonder if `concatenate_tractogram` should go in `dipy.tracking.utils` instead of `dipy.io.streamline`.

In general, it is not clear to me where all operations on Tractogram should go..... maybe a new module named `dipy.utils.streamline`

any opinion @frheault and @arokem?

**Comment by frheault** [comment]
Recently in Scilpy we added like 5 modules and changed the name and position of everything because we were tired of these challenges, so I believe it is a great question.

We created a tractogram module for operation at the tractogram level as opposed to the streamlines level. In this situation, I do not like utils (because everything is ending in there). So without a tractogram module, I would say streamline (but it is not great either)

**Comment by skoudoro** [comment]
Thank you for your review and feedback @frheault.

**Pull request merged**